lib/src/common/managed_stream_subscription.dart

```diff
@@ -22,22 +26,28 @@ class ManagedStreamSubscription<T> implements StreamSubscription<T> {
     _wrapOnError(onError);
   }
 
-  Future<Null> get didCancel => _didCancel.future;
+  Future<Null> get didComplete => _didComplete.future;
 
   @override
   bool get isPaused => _subscription.isPaused;
 
-  // TODO: Should we complete when the resulting future completes?
   @override
-  Future<E> asFuture<E>([E futureValue]) => _subscription.asFuture(futureValue);
+  Future<E> asFuture<E>([E futureValue]) {
+    return _subscription.asFuture(futureValue).then((value) {
+      _complete();
+      return value;
+    }).catchError((error, [stackTrace]) {
+      _complete();
```

Comment:
Wouldn't this silence an error if one was thrown from `_subscription.asFuture()`? I think this would be a perfect spot for `whenComplete()`:

```dart
return _subscription.asFuture(futureValue).whenComplete(_complete);
```